src/human_encoding/mod.rs

```diff
@@ -223,76 +223,108 @@ impl<J: Jet> Forest<J> {
 #[cfg(test)]
 mod tests {
     use crate::human_encoding::Forest;
-    use crate::jet::Core;
-    use crate::{Error, Value};
+    use crate::jet::{Core, Jet};
+    use crate::{BitMachine, Value};
     use std::collections::HashMap;
     use std::sync::Arc;
 
-    #[test]
-    fn to_witness_node_missing_witness() {
-        let s = "
-            wit1 := witness : 1 -> 2^32
-            wit2 := witness : 1 -> 2^32
-
-            wits_are_equal := comp (pair wit1 wit2) jet_eq_32 : 1 -> 2
-            main := comp wits_are_equal jet_verify            : 1 -> 1
-        ";
-        let forest = Forest::<Core>::parse(s).unwrap();
-        let mut witness = HashMap::new();
-        witness.insert(Arc::from("wit1"), Value::u32(1337));
+    fn assert_finalize_ok<J: Jet>(
+        s: &str,
+        witness: &HashMap<Arc<str>, Arc<Value>>,
+        env: &J::Environment,
+    ) {
+        let program = Forest::<J>::parse(s)
+            .expect("Failed to parse human encoding")
+            .to_witness_node(witness)
+            .expect("Forest is missing expected root")
+            .finalize()
+            .expect("Failed to finalize");
+        let mut mac = BitMachine::for_program(&program);
+        mac.exec(&program, env).expect("Failed to run program");
+    }
 
-        match forest.to_witness_node(&witness).unwrap().finalize() {
-            Ok(_) => panic!("Insufficient witness map should fail"),
-            Err(Error::IncompleteFinalization) => {}
-            Err(error) => panic!("Unexpected error {}", error),
+    fn assert_finalize_err<J: Jet>(
+        s: &str,
+        witness: &HashMap<Arc<str>, Arc<Value>>,
+        env: &J::Environment,
+        err_msg: &'static str,
+    ) {
+        let program = match Forest::<J>::parse(s)
+            .expect("Failed to parse human encoding")
+            .to_witness_node(witness)
+            .expect("Forest is missing expected root")
+            .finalize()
+        {
+            Ok(program) => program,
+            Err(error) => {
+                assert_eq!(error.to_string().as_str(), err_msg);
```

Comment:
In 60efcbf0a34204b28bc063804df609f523374c46:

You definitely do not need to write `to_string().as_str()` anywhere. Use `&error.to_string()`.